scene/gui/graph_element.cpp

```diff
@@ -0,0 +1,244 @@
+/**************************************************************************/
+
+#include "graph_element.h"
+
+#include "core/string/translation.h"
+#include "scene/gui/graph_edit.h"
+
+#ifdef TOOLS_ENABLED
+void GraphElement::_edit_set_position(const Point2 &p_position) {
+	GraphEdit *graph = Object::cast_to<GraphEdit>(get_parent());
+	if (graph) {
+		Point2 offset = (p_position + graph->get_scroll_offset()) * graph->get_zoom();
+		set_position_offset(offset);
+	}
+	set_position(p_position);
+}
+#endif
+
+void GraphElement::_close_requested() {
+	// Send focus to parent.
+	get_parent_control()->grab_focus();
+	emit_signal(SNAME("close_request"));
+}
+
+void GraphElement::_resort() {
+	Size2 size = get_size();
+
+	for (int i = 0; i < get_child_count(); i++) {
+		Control *child = Object::cast_to<Control>(get_child(i));
+		if (!child || !child->is_visible_in_tree()) {
+			continue;
+		}
+		if (child->is_set_as_top_level()) {
+			continue;
+		}
+
+		fit_child_in_rect(child, Rect2(Point2(), size));
+	}
+}
+
+Size2 GraphElement::get_minimum_size() const {
+	Size2 minsize;
+	for (int i = 0; i < get_child_count(); i++) {
+		Control *child = Object::cast_to<Control>(get_child(i));
+		if (!child) {
+			continue;
+		}
+		if (child->is_set_as_top_level()) {
+			continue;
+		}
+
+		Size2i size = child->get_combined_minimum_size();
+
+		minsize.x = MAX(minsize.x, size.x);
+		minsize.y += MAX(minsize.y, size.y);
+	}
+
+	return minsize;
+}
+
+void GraphElement::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_SORT_CHILDREN: {
+			_resort();
+		} break;
+
+		case NOTIFICATION_LAYOUT_DIRECTION_CHANGED:
+		case NOTIFICATION_TRANSLATION_CHANGED:
+		case NOTIFICATION_THEME_CHANGED: {
+			update_minimum_size();
+			queue_redraw();
+		} break;
+	}
+}
+
+void GraphElement::_validate_property(PropertyInfo &p_property) const {
+	Control::_validate_property(p_property);
+	GraphEdit *graph = Object::cast_to<GraphEdit>(get_parent());
+	if (graph) {
+		if (p_property.name == "position") {
+			p_property.usage |= PROPERTY_USAGE_READ_ONLY;
+		}
+	}
+}
+
+void GraphElement::set_position_offset(const Vector2 &p_offset) {
+	position_offset = p_offset;
+	emit_signal(SNAME("position_offset_changed"));
+	queue_redraw();
+}
+
+Vector2 GraphElement::get_position_offset() const {
+	return position_offset;
+}
+
+void GraphElement::set_selected(bool p_selected) {
+	if (!is_selectable() || selected == p_selected) {
+		return;
+	}
+	selected = p_selected;
+	emit_signal(p_selected ? SNAME("node_selected") : SNAME("node_deselected"));
+	queue_redraw();
+}
+
+bool GraphElement::is_selected() {
+	return selected;
+}
+
+void GraphElement::set_drag(bool p_drag) {
+	if (p_drag) {
+		drag_from = get_position_offset();
+	} else {
+		emit_signal(SNAME("dragged"), drag_from, get_position_offset()); // Required for undo/redo.
+	}
+}
+
+Vector2 GraphElement::get_drag_from() {
+	return drag_from;
+}
+
+void GraphElement::gui_input(const Ref<InputEvent> &p_ev) {
```

Comment:
This method is still present in `GraphNode` verbatim. Should it be removed from there?

Reply:
I think so, it can be added back in case we need to add specific behavior. Currently, the only difference is the error message (which will be a bit less precise when the GraphNode implementation of the method is removed, but that's ok IMO).